src/riak_kv_backend.erl

```diff
@@ -231,62 +206,55 @@ fold_objects({Backend, State}) ->
+                                           [],
+                                           State),
+                     lists:sort(Keys)
+                 end),
 
-             ?assertEqual([{{B1,K1}, <<"v1">>}],
-                          begin
-                              {ok, Objects1} =
-                                  Backend:fold_objects(FoldObjectsFun,
-                                                       [],
-                                                       [],
-                                                       State),
-                              lists:sort(ObjFilter(Objects1))
-                          end),
-             ?assertMatch({ok, _},
-                          Backend:put(B3, K3, [], <<"v3">>, State)),
-             ?assertEqual([{{B1,K1},<<"v1">>},
-                           {{B3,K3},<<"v3">>}],
-                          begin
-                              {ok, Objects} =
-                                  Backend:fold_objects(FoldObjectsFun,
-                                                       [],
-                                                       [],
-                                                       State),
-                              lists:sort(ObjFilter(Objects))
-                          end)
-     end
-    }.
+    ?assertEqual([{{B1,K1}, <<"v1">>}],
+                 begin
+                     {ok, Objects1} =
+                         Backend:fold_objects(FoldObjectsFun,
+                                              [],
+                                              [],
+                                              State),
+                     lists:sort(ObjFilter(Objects1))
+                 end),
+    {ok, State2} =  Backend:put(B3, K3, [], <<"v3">>, State),
+    ?assertEqual([{{B1,K1},<<"v1">>},
+                  {{B3,K3},<<"v3">>}],
+                 begin
+                     {ok, Objects} =
+                         Backend:fold_objects(FoldObjectsFun,
+                                              [],
+                                              [],
+                                              State2),
+                     lists:sort(ObjFilter(Objects))
+                 end),
+    empty_check(Backend, State).
 
-empty_check({Backend, State}) ->
+empty_check(Backend, State) ->
     {B1, _B2, K1, _K2} = make_bs_and_ks(Backend),
     B3 = make_test_bucket(Backend, 3),
     K3 = make_test_key(Backend, 3),
-    {"is_empty test",
-     fun() ->
-             ?assertEqual(false, Backend:is_empty(State)),
-             ?assertMatch({ok, _}, Backend:delete(B1,K1, [], State)),
-             ?assertMatch({ok, _}, Backend:delete(B3,K3, [], State)),
-             ?assertEqual(true, Backend:is_empty(State))
-     end
-    }.
+    ?assertEqual(false, Backend:is_empty(State)),
+    {ok, State2} = Backend:delete(B1,K1, [], State),
+    {ok, State3} = Backend:delete(B3,K3, [], State2),
+    ?assertEqual(true, Backend:is_empty(State3)),
```

Comment:
I don't know @slfritchie, I just made a basic functional backend (`dict`) and it failed all these tests as they each expect side effects in the backend and ignore the `State` var. I simply attempted to re-write them exactly as is, but use that `State` var so they work for purely functional backends.